src/algebraic_geometry/morphisms/basic.lean

```diff
@@ -0,0 +1,165 @@
+  le_inf := λ P₁ P₂ P₃ h₁ h₂ X Y f H, ⟨h₁ f H, h₂ f H⟩ }
+
+/-- A `affine_target_morphism_property` can be extended to a `morphism_property` such that it
+*never* holds when the target is not affine -/
+def affine_target_morphism_property.to_property (P : affine_target_morphism_property) :
+  morphism_property :=
+λ X Y f, ∃ h, @@P f h
+
+lemma affine_target_morphism_property.to_property_apply (P : affine_target_morphism_property)
+  {X Y : Scheme} (f : X ⟶ Y) [is_affine Y] :
+  P.to_property f ↔ P f := by { delta affine_target_morphism_property.to_property, simp [*] }
+
+/-- A morphism property `respects_iso` if it still holds when composed with an isomorphism -/
+def respects_iso (P : morphism_property) :=
+  (∀ {X Y Z} (e : X ≅ Y) (f : Y ⟶ Z), P f → P (e.hom ≫ f)) ∧
+  (∀ {X Y Z} (e : Y ≅ Z) (f : X ⟶ Y), P f → P (f ≫ e.hom))
+
+/-- A morphism property is `stable_under_composition` if the composition of two such morphisms
+still falls in the class. -/
+def stable_under_composition (P : morphism_property) :=
+  ∀ ⦃X Y Z⦄ (f : X ⟶ Y) (g : Y ⟶ Z), P f → P g → (P (f ≫ g))
+
+/-- A morphism property is `stable_under_composition` if the base change of such a morphism
+still falls in the class. -/
+def stable_under_base_change
+  (P : morphism_property) : Prop :=
+∀ ⦃X Y S : Scheme⦄ (f : X ⟶ S) (g : Y ⟶ S), P g → P (pullback.fst : pullback f g ⟶ X)
+
+lemma stable_under_composition.respects_iso {P : morphism_property}
+  (hP : stable_under_composition P) (hP' : ∀ {X Y} (e : X ≅ Y), P e.hom) : respects_iso P :=
+⟨λ X Y Z e f hf, hP _ _ (hP' e) hf, λ X Y Z e f hf, hP _ _ hf (hP' e)⟩
+
+lemma respects_iso.cancel_left_is_iso {P : morphism_property}
+  (hP : respects_iso P) {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [is_iso f] :
+    P (f ≫ g) ↔ P g :=
+⟨λ h, by simpa using hP.1 (as_iso f).symm (f ≫ g) h, hP.1 (as_iso f) g⟩
+
+lemma respects_iso.cancel_right_is_iso {P : morphism_property}
+  (hP : respects_iso P) {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [is_iso g] :
+    P (f ≫ g) ↔ P f :=
+⟨λ h, by simpa using hP.2 (as_iso g).symm (f ≫ g) h, hP.2 (as_iso g) f⟩
+
+lemma respects_iso.affine_cancel_left_is_iso {P : affine_target_morphism_property}
+  (hP : respects_iso P.to_property) {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [is_iso f]
+    [is_affine Z] : P (f ≫ g) ↔ P g :=
+by convert hP.cancel_left_is_iso f g; rw P.to_property_apply
+
+lemma respects_iso.affine_cancel_right_is_iso {P : affine_target_morphism_property}
+  (hP : respects_iso P.to_property) {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [is_iso g]
+  [is_affine Y] [is_affine Z] : P (f ≫ g) ↔ P f :=
+by convert hP.cancel_right_is_iso f g; rw P.to_property_apply
+
+/-- This is here to mirror `stable_under_base_change.snd`. -/
+@[nolint unused_arguments]
+lemma stable_under_base_change.fst {P : morphism_property}
+  (hP : stable_under_base_change P) (hP' : respects_iso P) {X Y S : Scheme} (f : X ⟶ S)
+  (g : Y ⟶ S) (H : P g) : P (pullback.fst : pullback f g ⟶ X) :=
+hP f g H
```

Comment:
It doesn't hurt if you just drop `hP'`?

Reply:
It is to mirror the type of `stable_under_base_change.snd`. It would be annoying if you needed to remember which is the one that needs `respects_iso`, and needs more work during copy-pasting when providing some symmetric results down the road.

Reply:
sounds reasonable